### ISSUE TITLE
Crystal 1.4.0, module can't overwrite instance variable

### ISSUE THREAD

**Issue by carlhoerberg** [crystal-lang/crystal#11994]
With Crystal 1.4.0:

```crystal
class A
  @x = false
end

module B
  @x = true
end

class C < A
  include B
end

C.new
```

```
✗ crystal run h.cr --error-trace
In h.cr:6:3

 6 | @x = true
     ^
Error: instance variable '@x' of C is already defined in A
```

The workaround:

```crystal
class A
  @x = false
end

module B
end

class C < A
  include B
  @x = true
end

C.new
```

compiles

**Comment by asterite** [comment]
This was introduced in https://github.com/crystal-lang/crystal/pull/11768

I really think we should revert that because it's a breaking change, as you can see in this issue. I can try to think of a proper fix, maybe for 1.5. But I'd rather not introduce breaking changes with incorrect fixes. This bug existed for some time now, I think it's fine to let it be for a little longer.

**Comment by asterite** [comment]
Actually, maybe it's fine if this errors. @carlhoerberg did this change break existing code, or did you find it in new code?

**Comment by carlhoerberg** [comment]
Ok, it broke existing code for us

**Comment by asterite** [comment]
Then I think we should undo the "fix"

**Comment by beta-ziliani** [comment]
@carlhoerberg that kind of re-definition of a variable in a class and a method was not really supported before 1.4.0 either. The fact that it worked was sheer luck: if you add an instance variable to `C`, say `@y = 1`, you'll reach an `Invalid memory access` (see #11673).

While it definitively deserves a discussion of if we want to allow such duplication of definitions, and how to interpret them, we shouldn't assume it was supported before 1.4.0. I would rather interpret it as 1.4.0 detecting a bug in existing code that happened to work by chance.

This said, I'm curious about the actual use case and the original intention.

Moving forward, I see the following options, each with its issues: 
1. Revert #11768 like @asterite proposed. Then, we get an awful, obscure dump that happens to work in one lucky corner case.
2. Detect the corner case and make it work in just that case (including class `C` having no instance variables of its own). Then, we get a typechecker that is way too ad-hoc-y until a proper solution is implemented.
3. Solve the issue for real. This requires a proper discussion and implementation, that is not going to be resolved before 1.4.1 for sure, and that not necessarily will keep the original code working.
